var/spack/repos/builtin/packages/zlib/package.py

```diff
@@ -25,27 +25,42 @@ class Zlib(Package):
             description='Produce position-independent code (for shared libs)')
     variant('shared', default=True,
             description='Enables the build of shared libraries.')
+    variant('debug',   default=False,
+            description='Build debug code (-g -O0)')
     variant('optimize', default=True,
             description='Enable -O2 for a more optimized lib')
 
+    conflicts('+debug+optimize')
+    conflicts('+shared', when='@:1.2.5')
+
     patch('w_patch.patch', when="@1.2.11%cce")
 
+    def patch(self):
+        if self.spec.satisfies('@:1.2.5'):
+            filter_file(r'stdio.h', 'stdio.h>\n#include <errno.h', 'gzio.c')
+
     @property
     def libs(self):
         shared = '+shared' in self.spec
         return find_libraries(
             ['libz'], root=self.prefix, recursive=True, shared=shared
         )
 
-    def setup_build_environment(self, env):
-        if '+pic' in self.spec:
-            env.append_flags('CFLAGS', self.compiler.cc_pic_flag)
-        if '+optimize' in self.spec:
-            env.append_flags('CFLAGS', '-O2')
+    def flag_handler(self, name, flags):
+        if name in ('cflags'):
+            if '+pic' in self.spec:
+                flags.append(self.compiler.cc_pic_flag)
+            if '+debug' in self.spec:
+                flags.append('-g -O0')
+                flags = list(map(  # Fix to change all -O to -O0
+                    lambda x: "-O0" if x.startswith('-O') else x, flags))
+            if '+optimize' in self.spec:
+                flags.append('-O2')  # TODO: Really only if no -O already set
```

Comment:
```suggestion
                flags.append('-O2')  # TODO: Don't add -O2 if -O was already set elsewhere.
```
A lot of comments in this repo are somewhat mysterious to me, and I think the current comment might not be understandable to many readers (I'm thinking of people who learned English as a second language). I received some advice to use complete sentences a while ago and I have found it immensely useful.

Reply:
Also, if we have two conditional branches (with `+debug` and with `+optimize`) which depend upon whether any optimization flags have been set, would it make sense to process them together? Something like:

```python
is_optimize = '+optimize' in self.spec
is_debug = '+debug' in self.spec
if is_optimize or is_debug:
    flags = [arg for arg in flags if not arg.startswith('-O')]
if is_optimize:
    flags.append('-O2')
if is_debug:
    flags.append('-g -O0')
```

Reply:
Going to mark this as resolved since I think there is a better answer to this in modifying the `flag_handler()` API.